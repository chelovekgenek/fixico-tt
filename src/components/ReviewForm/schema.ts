import * as Yup from "yup";
import { IReviewDraft } from "store/entities/reviews/types";

export const schema = Yup.object().shape<IReviewDraft>({
  body: Yup.string()
    .min(2)
    .max(32)
    .required(),
  rating: Yup.number()
    .min(0)
    .max(5)
    .required()
});

export const initialValues: Yup.InferType<typeof schema> = {
  body: "",
  rating: 0
};
