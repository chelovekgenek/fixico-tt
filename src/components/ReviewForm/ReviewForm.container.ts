import { connect } from "react-redux";
import { withFormik } from "formik";

import { TAppState } from "store/entities";

import { schema, initialValues } from "./schema";
import { IProps, ReviewForm } from "./ReviewForm";
import { ReviewsPostActions } from "store/entities/reviews";

export interface IDispatchProps {
  create: typeof ReviewsPostActions.request;
}

export const ReviewFormContainer = connect<{}, IDispatchProps, {}, TAppState>(
  null,
  {
    create: ReviewsPostActions.request
  }
)(
  withFormik<IProps, typeof initialValues>({
    handleSubmit: (values, { props, resetForm }) => {
      props.create(values);
      resetForm();
    },
    validationSchema: schema,
    validateOnBlur: true
  })(ReviewForm)
);
