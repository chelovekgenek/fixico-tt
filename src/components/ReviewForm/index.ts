export { ReviewFormContainer as ReviewForm } from "./ReviewForm.container";
