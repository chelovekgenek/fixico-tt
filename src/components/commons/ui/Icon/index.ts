export { Icon } from "@material-ui/core";
