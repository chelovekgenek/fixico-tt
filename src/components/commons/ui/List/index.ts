import * as MUI from "@material-ui/core";

import { Text } from "./Item";
import { Avatar } from "./Avatar";

export const List = Object.assign(MUI.List, {
  Item: Object.assign(MUI.ListItem, {
    Text,
    Avatar,
    SecondaryAction: MUI.ListItemSecondaryAction
  })
});
