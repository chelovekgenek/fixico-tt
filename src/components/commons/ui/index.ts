export * from "./Button";
export * from "./IconButton";
export * from "./Icon";
export * from "./List";
export * from "./Pagination";
export * from "./Rating";
