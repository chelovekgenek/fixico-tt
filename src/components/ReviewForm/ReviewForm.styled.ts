import styled from "styled-components";

import { Form } from "formik";
import { Grid } from "components/commons/layout";

export const FormWrapper = styled(Form)`
  margin: 1rem 0 2rem;
`;

export const GridCenterItem = styled(Grid)`
  display: flex;
  align-items: center;
`;
