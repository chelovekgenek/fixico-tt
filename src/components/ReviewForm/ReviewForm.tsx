import React, { useCallback } from "react";
import { FormikProps } from "formik";

import { Grid } from "components/commons/layout";
import { Input, Rating } from "components/commons/form";
import { Button } from "components/commons/ui";

import { initialValues } from "./schema";
import { IDispatchProps } from "./ReviewForm.container";
import { FormWrapper, GridCenterItem } from "./ReviewForm.styled";

export interface IProps extends IDispatchProps {}

export const ReviewForm: React.FC<
  IProps & FormikProps<typeof initialValues>
> = ({ handleSubmit, isValid }) => {
  const handleClicked = useCallback(() => handleSubmit(), [handleSubmit]);
  return (
    <FormWrapper>
      <Grid container justify="space-between">
        <GridCenterItem item>
          <Input label="Body" name="body" />
        </GridCenterItem>
        <GridCenterItem item>
          <Rating label="Rating" name="rating" />
        </GridCenterItem>
        <GridCenterItem>
          <Button
            variant="contained"
            color="primary"
            onClick={handleClicked}
            disabled={!isValid}
          >
            Create
          </Button>
        </GridCenterItem>
      </Grid>
    </FormWrapper>
  );
};
