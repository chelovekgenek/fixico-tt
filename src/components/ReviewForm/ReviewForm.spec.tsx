import React from "react";
import { shallow } from "enzyme";

import formProps from "mocks/formik-form-props";
import { Button } from "components/commons/ui";

import { ReviewForm } from "./ReviewForm";
import { initialValues } from "./schema";

describe("ReviewForm", () => {
  const render = (
    props: Partial<React.ComponentProps<typeof ReviewForm>> = {}
  ) =>
    shallow(
      <ReviewForm
        {...formProps}
        initialValues={initialValues}
        values={initialValues}
        create={jest.fn() as any}
        {...props}
      />
    );
  it("should render", () => {
    expect(render()).toMatchSnapshot();
  });
  it("should unlock submit button when prop is passed", () => {
    expect(
      render({ isValid: true })
        .find(Button)
        .prop("disabled")
    ).toBe(false);
  });
  it("should call Submit when button is clicked", () => {
    render({ isValid: true })
      .find(Button)
      .simulate("click");
    expect(formProps.handleSubmit).toBeCalled();
  });
});
