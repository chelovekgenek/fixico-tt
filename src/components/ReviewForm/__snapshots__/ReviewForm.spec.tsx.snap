// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewForm should render 1`] = `
<Styled(Form)>
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="space-between"
  >
    <Styled(WithStyles(ForwardRef(Grid)))
      item={true}
    >
      <Input
        label="Body"
        name="body"
      />
    </Styled(WithStyles(ForwardRef(Grid)))>
    <Styled(WithStyles(ForwardRef(Grid)))
      item={true}
    >
      <Rating
        label="Rating"
        name="rating"
      />
    </Styled(WithStyles(ForwardRef(Grid)))>
    <Styled(WithStyles(ForwardRef(Grid)))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={true}
        onClick={[Function]}
        variant="contained"
      >
        Create
      </WithStyles(ForwardRef(Button))>
    </Styled(WithStyles(ForwardRef(Grid)))>
  </WithStyles(ForwardRef(Grid))>
</Styled(Form)>
`;
